File: core/java/src/main/java/hara/lang/protocol/IToPersistent.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

@HaraProtocolBinding(
    namespace = "std.protocol.itopersistent", name = "IToPersistent", parents = {"IMutable"})
public interface IToPersistent extends IMutable {
  @HaraMethod(value = "to-persistent", arity = 1)
  IPersistent toPersistent();
}

File: core/rust/src/lang/protocol/itopersistent.rs
use super::IMutable;
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.itopersistent",
    name = "IToPersistent",
    parents = ["IMutable"]
)]
pub trait IToPersistent: IMutable {
    type Persistent;

    #[hara_method(value = "to-persistent", arity = 1)]
    fn to_persistent(&mut self) -> Self::Persistent;
}
